add makefile to requirements

badGarnet opened the pull request:
This PR resolves #1294 by adding a Makefile to compile requirements. This makefile respects the dependencies between file and will compile them in order. E.g., extra-*.txt will be compiled __after__ base.txt is updated.

Test locally by simply running `make pip-compile` or `cd requirements && make clean && make all`

cragwolfe commented:
why this is better than the existing solution? in the off chance that the existing `make pip-compile` results in inconsistent dependencies, this would get caught by https://github.com/Unstructured-IO/unstructured/blob/b710baf/.github/workflows/ci.yml#L70 .

but as long as pins are being kept in `constraints.in`, this really shouldn't happen in the first place.

badGarnet commented:
> why this is better than the existing solution? in the off chance that the existing `make pip-compile` results in inconsistent dependencies, this would get caught by https://github.com/Unstructured-IO/unstructured/blob/b710baf/.github/workflows/ci.yml#L70 .
> 
> but as long as pins are being kept in `constraints.in`, this really shouldn't happen in the first place.

The current method relies on `base.txt` is compiled first because of alphabetical order. The other extras that depend on `base.txt` are able to get updated `base.txt` because of naming. Additionally if we have more layered dependencies like `dev.txt` we may not be able to rely on the file names.

Makefile is designed to handle compilation dependency so I would argue that is the safest way to ensure pip-compile runs correctly. Now I do concede that with current requirements this may seem a bit over engineering.

badGarnet commented:
probably not needed for now

rbiseck3 commented:
Got this:
```shell
❯ make pip-compile
scripts/pip-compile.sh: line 11: MAKE[1]:: command not found
cp: requirements/build.txt: No such file or directory
make: *** [pip-compile] Error 1
```

rbiseck3 commented on scripts/pip-compile.sh:
```diff
@@ -8,5 +8,5 @@ if ! python -c "import sys; assert sys.version_info.major == $major and sys.vers
 fi
 
-cd ./requirements && $(MAKE) *.txt
+cd ./requirements && make *.txt && cd ..
```
Can use `pushd`/`popd` to navigate to/from the requirements dir. Has some pleasantries: [Navigating with popd](https://opensource.com/article/19/8/navigating-bash-shell-pushd-popd#navigating-with-popd)

qued commented:
Also there used to be logic to copy the `build.txt` file in `requirements/` into the `docs/` folder overwriting `docs/requirements.txt`. In other words, the file `build.txt` is really the requirements file for `sphinx` builds. I think that logic is missing from what we are currently doing as well.

badGarnet commented:
> Also there used to be logic to copy the `build.txt` file in `requirements/` into the `docs/` folder overwriting `docs/requirements.txt`. In other words, the file `build.txt` is really the requirements file for `sphinx` builds. I think that logic is missing from what we are currently doing as well.

updated makefile to work on subdirectories and added back the copy line

qued reviewed (approved):
Overall this LGTM!

Two other non-blocking nice-to-have things I'd flag are:
1. It looks like the formatting of the output files is slightly different (see attached image) so maybe we want to do a full re-compile of everything to put it in the new format? <img width="863" alt="image" src="https://github.com/Unstructured-IO/unstructured/assets/64741807/f580a639-4f3c-4ae1-8d95-77318bc8ef54">
2. Speaking of doing a full re-compile, I'm not totally sure how it's being determined when files need to be updated, but sometimes they're skipped (see image). I'm not sure it can make the determination that nothing needs to be done when new versions could have been released to pypi. But even assuming the logic is fine, it would be nice to have a way to force a compile on everything. 
<img width="314" alt="image" src="https://github.com/Unstructured-IO/unstructured/assets/64741807/9e4bcf2f-f4e4-475a-8fe1-72e63c277852">

badGarnet commented:
> Overall this LGTM!
> 
> Two other non-blocking nice-to-have things I'd flag are:
> 
> 1. It looks like the formatting of the output files is slightly different (see attached image) so maybe we want to do a full re-compile of everything to put it in the new format? <img alt="image" width="863" src="https://user-images.githubusercontent.com/64741807/279137060-f580a639-4f3c-4ae1-8d95-77318bc8ef54.png">
> 2. Speaking of doing a full re-compile, I'm not totally sure how it's being determined when files need to be updated, but sometimes they're skipped (see image). I'm not sure it can make the determination that nothing needs to be done when new versions could have been released to pypi. But even assuming the logic is fine, it would be nice to have a way to force a compile on everything.
> 
> <img alt="image" width="314" src="https://user-images.githubusercontent.com/64741807/279143422-9e4bcf2f-f4e4-475a-8fe1-72e63c277852.png">

1. this is due to where the pip compile command is executed; using the make file it is executed inside `requirements` folder
2. I can add a `make clean` to remove `.txt` then in script we can do `make clean && make all` to force recompile everything.
cc @rbiseck3 @ryannikolaidis
